manual: document modular explicits

gasche opened the pull request:
This is a companion PR to https://github.com/ocaml/ocaml/pull/13275 , which implement Modular Explicits, and is currently under review. The present PRs documents the feature in the manual. I adopted a commit by @samsa1 from #13275 (which extends the reference description of first-class modules to also cover module-dependent functions), and added a substantial amount of new content that tries to provide a readable introduction of how to use module-dependent functions and first-class modules, in a future world where modular explicits have been merged.

I think that it is worth sending this documentation PR in advance, because it can typically be reviewed by other people than #13275, which may be interested in how we describe the feature or opinions on how to do it better.

In 5.3, the [first-class modules section](https://ocaml.org/manual/5.3/firstclassmodules.html) of the "Languages Extension" chapter starts with the usual hard-to-read "reference" part with grammar and minimal mentions of what the grammar does, and is then followed by some nice examples. I decided to reorder things with first an "Introduction" section that tries to explain what the two features are, in an example-based way (reusing the previous examples), and then the obscure reference stuff.

Following a recommendation by @Octachron, I present module-dependent functions first as possibly the most common/natural construct, and then the full expressivity of first-class modules (with dynamic module selection) only second. I try to explain the restrictions/limitations around the two features, and have a short paragraph on how things were before the modular-explicits merge.

samsa1 committed:
Modified manual for explain modular explicits

gasche committed:
manual: readable 'introduction' in the modular-explicits extension chapter

Reviewed-by: Samuel Vivien <samuel.vivien@inria.fr>
Reviewed-by: Shon Feder <shon.feder@gmail.com>
Reviewed-by: Jules Aguillon <jules@j3s.fr>
Reviewed-by: Ali Caglayan <alizter@gmail.com>
Reviewed-by: Richard Eisenberg <reisenberg@janestreet.com>
Reviewed-by: KC Sivaramakrishnan <kc@kcsrk.info>
Reviewed-by: Didier Remy <didier.remy@inria.fr>
Reviewed-by: François Pottier <francois.pottier@inria.fr>

nojb commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
-makes it possible to pack a module as a first-class value, which can
-later be dynamically unpacked into a module.
+First-class modules and module-dependent functions make it possible to
+manipulate functions as values, in particular to pass them as function
```
```suggestion
manipulate modules as values, in particular to pass them as function
```

kayceesrk replied:
s/functions as values/modules as values

yotambarnoy commented:
Is the mapM example purely negative i.e. it's just an illustration of what cannot work?

gasche commented:
The `mapM` example works with modular explicits, but not with first-class modules only. (It can also be expressed using a functor). I can clarify this.

xvw commented:
> The `mapM` example works with modular explicits, but not with first-class modules only. (It can also be expressed using a functor). I can clarify this.

Although I think the `mapM` example is cool, I think the name `traverse` is more meaningful.

yotambarnoy commented:
I believe mapM already exists in the haskell lingo.

xvw commented:
Yes, mainly for reasons of class overlapping (and `traverse` also exists). Doesn't the fact that `fmap` exists in the Haskell lingo does not mean that it has to be integrated into teaching (FMPOV).

gasche commented:
I see some regularity in functions called `fooM`, where `foo` takes a higher-order function `(bar... -> baz)` and `fooM` a higher-order function `(bar... -> baz M.t)` -- for example we can also have `filterM`, `for_allM`, etc. This being said, if a consensus forms that `traverse` is less scary for non-monadic programmers, I'm okay with switching to that. (I guess that Haskell people would then point out that we don't need a whole monad for traverse, a traversable functor is enough).

samsa1 commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
+
+To apply the "sort" function, we use the construction "(module M : S)"
+to provide the argument, which can typically be written "(module M)"
+as the signature can be inferred from the function.
```
I'm not convinced by this sentence. I tend to consider that for module-dependent functions `(module M)` is the default and `(module M : S)` is just syntactic sugar for `(module (M : S))`. This must be opposed to first-class modules where `(module M : S)` means `(module M) : (module S)` which is drastically different.

gasche replied:
Here is the mental model that I would like to offer to users:

1. There is a general approach of having term-level code parametrized over modules, here is the syntax and some examples.
2. When you look at the details, there are two different mechanisms (actually-dependent functions vs. first-class modules) which are non-comporable in expressivity.

I would like to write as much of the introduction as possible with (1) in mind, in a way that is reasonably agnostic to which of the two mechanisms can be used in a particular example.

In this setting my view is that `(module M : S)` is the correct construction to emphasize: it is unambiguous, and provides the most precise information for both mechanisms.

diremy replied:
I would say `can be inferred from the function type.` rather than `can be inferred from the function.`

Octachron replied:
I read @samsa1's comment as being more concerned that on the application side  `(module M)` is always sufficient for module-dependent function. Contrarily `(module M:S)` (aka `(module (M:S))` is only useful to *reject* more programs.  In other words, I would suggest to simplify the previous sentence to:

> To apply the "sort" function, we use the construction "(module M)" to provide the argument:

gasche replied:
> To apply the "sort" function, we use the construction "(module M)" to provide the argument, or "(module M : S)" to be explicit about the intended signature.

samsa1 commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
+let example = sort (module ISetDecr) [3; 1; 2]
+
+(* Note: any module path is accepted, including functor applications *)
+let example = sort (module Set.Make(Int)) [3; 1; 2]
```
```ocaml
let example = sort (module Set.Make(struct type t = int let compare = Int.compare end))
```
is also accepted while
```ocaml
module F () : Set.S = Set.Make(Int)
let example = sort (module F())
```
is rejected

gasche replied:
I don't know why `Make(struct .. end)` is accepted (it looks a bit fishy to me!), but I just removed the comment to avoid the difficulty. Thanks!

samsa1 replied:
The reason is that `sort (module M)` is of type `M.elt list -> M.elt list` and as the typechecker is able to inline the definition of `M.elt` (ie : `int`) then the return type is valid. As such it is accepted because this does not pose any soundness issue.

samsa1 commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +138,89 @@ let draw_using_device device_name picture =
+    (match sort_order with
+     | `Incr -> (module ISetIncr)
+     | `Decr -> (module ISetDecr))
+    [2; 4]
```
Maybe also add a version of this example using using `val` to allow keeping a module-dependent function ?
I feel that the presentation here tends to hint that changing the definition of the function `sort` is the way to go but it is not always possible (for example if you didn't define the function or if it uses a parametrized type).

gasche replied:
I'm not sure which `val` version you have in mind, can you write it down?

samsa1 replied:
The following example is valid and seems like a small diff from the invalid example in the documentation.
```ocaml
let valid =
  sort
    (module (val (match sort_order with
     | `Incr -> (module ISetIncr)
     | `Decr -> (module ISetDecr)) : Set.S with type elt = int))
    [2; 4]
```

gasche replied:
I'm surprised that this works and it means that there is something I don't understand. If for any `t : (module S)` and module-dependent function `f : (module M : S) -> ...` we could rewrite `f t` into `f (module (val t))`, we would not forbid non-static applications. So when does this rewrite not work? (My guess would be that this related to the ability to add equalities to `S`, which only work for non-parametrized types.)

samsa1 replied:
If we have `f : (module M : S) -> C[M.elt]`, when computing the type of `f (module ME)` we have solutions :
- extract a path `P` from `ME` to have the return type `C[P.elt]`
- inline the definition of the field `elt` in `ME` (ie : t) to have the return type `C[t]`.

The example given above uses the second approach. The reason why we don't allow first-class modules here is because `(module S)` and `(module S with type elt = ...)` are not compatible. So :
- either the module-dependent function takes an argument of the form `(module S with type elt = ...)` and we can remove the dependency.
- or we are in the form `(module S)` which means that we won't be able to inline the definition of the field `elt`.

gasche replied:
I pushed a modified version of this discussion that takes @samsa1's feedback into account, and he reviewed it in-person.

samsa1 reviewed (commented):
This new documentation is much better than initial one I proposed so I am strongly in favor of this PR.
I did a first pass-through for sanity checking and have a few comments.

The biggest one would be the lack of explanation of the compatibility between first-class-modules and module dependent functions. ie :
- Unification between both that happens dynamically,
- Ground coercion

shonfeder commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +138,89 @@ let draw_using_device device_name picture =
+    [2; 4]
+\end{caml_example}
+
+This approach works whenever the types used in the right-hand-size of
```
Is this meant to be?

```suggestion
This approach works whenever the types used in the right-hand-side of
```

gasche replied:
Fixed, thanks! (I haven't updated the remote/PR branch yet.)

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
-          'type' typeconstr '=' typexpr
-;
-\end{syntax}
+Module-dependent functions were introduced in 5.5.)
```
Shouldn't this be `5.05` by the above convention?

gasche replied:
We used 4.0x numbers but then we decided to use 5.x directly instead.

Julow commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -102,8 +117,10 @@ module PDF = struct let draw () = () [@@ellipsis] end
-We can then select one implementation based on command-line
-arguments, for instance:
+The module expression "(val e : S)" can then \emph{unpack}
+a packed-module value "(e : module S)" back into a module. In our
```
Valid syntax is `(e : (module S))`.

gasche replied:
... no idea why we don't also support `(e : module S)`, but oh well... Thanks!

gasche committed:
manual: document a corner case found by Florian

gasche commented:
I modified the new introduction/tutorial part of this PR to address a non-trivial comment by @samsa1, and also to explain a potentially surprising example of code that fails to type check reported by @Octachron in https://github.com/ocaml/ocaml/pull/13275#discussion_r2101093278 .

Julow commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
-@'let' 'module' module-name '=' '(' "val" expr_1 ":" package-type ')'
- "in" expr_2@.
+\begin{caml_example}{verbatim}
+let sort (module Set : Set.S) li =
+  Set.elements (List.fold_right Set.add li Set.empty)
```
I did not expect the sort function to be parametrized over a `Set` but rather a `OrderedType`:
```ocaml
let sort (module OrderedType : Set.OrderedType) li =
  let module Set = Set.Make(OrderedType) in
  Set.elements (List.fold_right Set.add li Set.empty)
```
I think that's how the similar feature is used in other languages. It's also more consistent with the intention of letting the caller choose another comparison function.

I think parameterizing the `Set` was not needed for the rest of the explanation. For example, the Incr/Decr example becomes simpler with defining two different `OrderedType` rather than two different `Set`.

gasche replied:
Thanks for the feedback.

One reason to parametrize on `Set.S` is that it lets people provide non-uniform implementations, for example maybe we have a type `BitSet` that stores a set of small integers with a completely different implementation.

I think that both choices are sensible. I'm not sure what reads best in the manual. Maybe I'll try with your proposal to see how it turns out.

gasche replied:
I thought more about this, and I think that the current approach is fine. There is a continuum between abstracting over just a comparison function, abstracting on an OrderedType module, and abstracting over a complete Set.S implementation. I think that all are reasonable approaches but the one used in the tutorial has the advantage of (being slightly more expressive and) leading to short code -- it's a detail but `Set.S` is shorter than `Set.OrderedType`, and the different is not recouped by writing `S.elt` instead of `M.t`.

didierremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
+package type inference introduced in 4.00; structural comparison of
+package types introduced in 4.02.; fewer parens required starting from
+4.05.
 
```
There should be not line break here

gasche replied:
I think it is nice to distinguish visually between first-class modules and module-dependent functors, but I turned this into a `\\` to avoid a new paragraph.

didierremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
-Modules are typically thought of as static components. This extension
-makes it possible to pack a module as a first-class value, which can
-later be dynamically unpacked into a module.
+First-class modules and module-dependent functions make it possible to
```
Should be "Module-dependent functions and first-class modules" as in the section title.

didierremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
-@'let' 'module' module-name '=' '(' "val" expr_1 ":" package-type ')'
- "in" expr_2@.
+\begin{caml_example}{verbatim}
+let sort (module Set : Set.S) li =
```
Here and below the first/bound occurrence of  `Set` should be renamed, e.g. into `Myset` (or `M` or `X`) so that it is clearly a binding occurrence that has nothing to do with the existing `Set` module.  All used occurrences should be renamed accordingly. The very same renaming should also be applied to all similar variants of `sort` below.

gasche replied:
I renamed the argument into `MSet`, which I find more readable than `M` or `S`. Thanks!

didierremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
+let example = sort (module Set.Make(Int)) [3; 1; 2]
+\end{caml_example}
+
+This provides a syntactically lightweight alternative to functors:
```
(1) I would say "Here, this provides" not to imply that this is a general pattern that always applies, as the general encoding is not always that simple.
(2) I would just say "a lightweight" alternative instead of a "syntactically lightweight alternative".

didierremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +138,146 @@ let draw_using_device device_name picture =
+
+To avoid breaking type soundness, there is a restriction on functions
+that take module parameters "fun (module A : S) -> ...": if the type
+of the rest of the function (the "...") \emph{depends} on the module
```
I would say "the body of the function" instead of "the rest of the function".

didierremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +138,146 @@ let draw_using_device device_name picture =
+To avoid breaking type soundness, there is a restriction on functions
+that take module parameters "fun (module A : S) -> ...": if the type
+of the rest of the function (the "...") \emph{depends} on the module
+argument "A", then this function is only allowed to take static module
```
I am not sure that this is a good use of "static", since the underlying module may still be passed at runtime.
Why not say "explicit module arguments" instead of "static module arguments".
By opposition with "first-class module expressions" ?

gasche replied:
"Static module argument/application" is the terminology proposed by @samsa1 and I am fine with it. I don't think that we can use "explicit module argument", as to me this is clearly in contrast to "implicit module argument" (for a modular-implicit function) and not about being directly a module or a general expression.

samsa1 replied:
I don't have a strong opinion put it is the vocabulary currently used in the error messages and I such it should be coherent.

I also think @clementblaudeau might have an opinion here such that we have a coherence with other module related features

didierremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +138,146 @@ let draw_using_device device_name picture =
+function definition and not just the call-site.
+
+\begin{caml_example}{verbatim}
+let sort_nondep (type a) (module Set : Set.S with type elt = a) (li : a list) : a list =
```
Line too long...
```suggestion
let sort_nondep (type a) (module Set : Set.S with type elt = a) (li : a list) 
    : a list =
```

goldfirere commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
-It is not allowed in toplevel let bindings.
-Again @package-type@ can be omitted if it can be inferred from the
-enclosing pattern.
+One can write a function "sort0" that sorts element by adding then into
```
```suggestion
One can write a function "sort0" that sorts elements by adding them into
```

goldfirere commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
-are compared using the structural comparison of module types.
+Using module-dependent functions, it is also possible to express
+"sort" as a function parametrized over an ordered-set module; this
+lets caller choose their ordered-set implementation, which may be
```
```suggestion
lets the caller choose their ordered-set implementation, which may be
```

goldfirere commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
+\end{caml_example}
+
+This provides a syntactically lightweight alternative to functors:
+indeed, this "sort" function could also be expressed as follows:
```
```suggestion
This provides a syntactically lightweight alternative to functors.
Indeed, this "sort" function could also be expressed as follows:
```

goldfirere commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,107 @@
+
+\lparagraph{p:fst-mod-intro-dynamic}{First-class modules}
+
+In the function application "sort (module ISetIncr)" of our example
```
```suggestion
In the function application "sort (module ISetIncr)" from our example
```

goldfirere commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -231,3 +365,83 @@ let rec to_string: 'a. 'a Typ.typ -> 'a -> string =
+
+The expression @'(' 'module' module-expr ':' package-type ')'@ converts the
+module (structure or functor) denoted by module expression @module-expr@
+to a value of the core language that encapsulates this module.  The
```
```suggestion
to a value of the core language that packs this module.  The
```

goldfirere reviewed (commented):
Looks good -- I like this explanation!

gasche commented:
Thanks @goldfirere and @kayceesrk, I took your comments into account and rebased.

(My understanding of the present PR is that it is de-facto approved, in that it is waiting for the module-dependent functions *implementation* #13275 to be merged and we will merge it at the same time.)

fpottier commented:
I am interested in reading this documentation, but I was not able to build it; checking out this PR and attempting to build the manual results in a type error. Could you post a link to a rendered version, or post instructions to check out and build the document?

gasche commented:
A rendered version of the manual with this PR is available at <https://www.irif.fr/~scherer/tmp/modular-explicits-manual.pdf>.

(The reason this is hard to build is that the manual PR must be based on the modular-explicits PR to be able to run the examples in the manual, but I artifically rebased it on trunk for submission on github, otherwise the diff wrt. trunk would be too large and people could not review it easily through the github interface.)

Alizter commented:
> A rendered version of the manual with this PR is available at [irif.fr/~scherer/tmp/modular-explicits-manual.pdf](https://www.irif.fr/~scherer/tmp/modular-explicits-manual.pdf).

For others that are looking for it, the relevant section is 12.5.

fpottier commented:
Thanks @gasche . Here are some remarks about the new section 12.5.1 of the manual. I hope this can be useful. 

I would suggest writing "parameterized" instead of "parametrized".

"Functions parametrized over a module": I find it confusing that the first example, `sort0`, does *not* involve a function that is parameterized with a module, despite the title. Furthermore, `sort0` is arguably an example of something that should be avoided in OCaml, namely, using the function `Stdlib.compare` at an unknown type `a`. I believe that the text should make it clear that `sort0` is an example of a bad style that was sometimes used before module-dependent functions existed, and that this example can now be written in a better way using module-dependent functions.

In the description of `sort`, the argument is named sometimes `MSet`, sometimes `Set`. This confusion could be avoided.

It may be useful to explicitly show the results of `sort (module ISetIncr) [3; 1; 2]` and `sort (module ISetDecr) [3; 1; 2]`.

"any module path is accepted, including functor applications": intuitively, it seems to me that this makes sense only if the functor is applicative.

"First-class modules let you build dynamic values": why is the word `dynamic` typeset in typewriter font?

"To avoid breaking type soundness, there is a restriction [...]". I would suggest starting this sentence directly with "There is a restriction [...]". If you wish to explain why this restriction is necessary, do so later.

"there is a restriction on functions that take module parameters": replace with "there is a restriction on module-dependent functions". Remove "if the type of the body of the function (the ...) depends on the module argument A". Remove "We call them module-dependent functions". This terminology has already been introduced.

In "Paragraph ??", broken reference. Where is the function `process` defined? I cannot find it anywhere.

"Its argument here is not static, so the type-checker tried instead to change the function type to be non-dependent". Wow, this is a long and complex error message, which looks like a *type error* message, whereas the previous paragraph has just explained that applications of module-dependent functions are subject to a *purely syntactic* restriction, namely, the actual argument should have the syntactic form `(module M : S)`. So, either this error message is over-complicated, or (as seems to be the case here) the true rule is less restrictive than what has been suggested earlier: if the actual argument is not of the syntactic form `(module M : S)` then the type-checker attempts to weaken the module-dependent function type into a non-dependent type. (I can imagine that type inference complicates things: it is not necessarily known ahead of time whether the function is or is not module-dependent.) Anyway, in this example where it seems clear that the function type is known, is module-dependent, and cannot be meaningfully weakened to a non-dependent type, one would expect to see a much simpler error message, saying that the argument of a module-dependent function must have the syntactic form `(module M : S)`.

"Note that removing dependencies on a type within a module can only be done for non-parametrized type like elt". It may be unclear to the reader what is meant by "removing dependencies on a type within a module". I suppose that it refers to the technique that has just been illustrated, but this is not crystal clear. Also, you could say more explicitly why this technique works only for base types, not for parameterized types: the reason is that OCaml does not have higher kinds (right?). I mean, the syntax `(type a)` cannot be used to introduce a parameterized local abstract type.

In my mind, the discussion of `sort_nondep` should be relegated to an explicitly-delimited remark, because it is of secondary interest; it discusses a work-around that used to be useful before module-dependent functions were available, but is no longer so useful.

Conversely, the example of `Monad` and `mapM`, which is currently presented as a situation where the (now-useless) work-around does not work, is interesting in its own right and should be presented as a typical example of a module-dependent function, without any reference to `sort_nondep` or to "removing dependencies on a type within a module".

I would suggest extending the example of `mapM` with an example of the use of `mapM` with a concrete monad.

gasche commented:
Excellent feedback, thanks!

> In my mind, the discussion of `sort_nondep` should be relegated to an explicitly-delimited remark, because it is of secondary interest; it discusses a work-around that used to be useful before module-dependent functions were available, but is no longer so useful.

I would still want to keep it reasonably visible because first-class modules have been around for a while, so I believe that this programming idiom (`(type a) (module M : S with type t = a)`) is wide-spread, so people will keep encountering it for a while, and also non-obvious, so they need documentation.

> Wow, this is a long and complex error message, which looks like a type error message, whereas the previous paragraph has just explained that applications of module-dependent functions are subject to a purely syntactic restriction [...] If the actual argument is not of the syntactic form (module M : S) then the type-checker attempts to weaken the module-dependent function type into a non-dependent type.

Yes, this is broadly what is going on: when the type-checker detects that dependency is not supported, then it pessimizes the type-checking rules to a non-dependent application. The problem is that the syntactic form of the argument is not the only reason why pessimization would be necessary, and the type-checker cannot tell when there is an error if the pessimization was intended by the programmer or part of the programming error.

> Anyway, in this example where it seems clear that the function type is known, is module-dependent, and cannot be meaningfully weakened to a non-dependent type, one would expect to see a much simpler error message, saying that the argument of a module-dependent function must have the syntactic form (module M : S).

If the error message said too confidently that the problem is with the syntactic form of the argument, it could lead users in a completely wrong direction because maybe actualy their problem is different.

The root cause of this hesitation in the error message is the ambiguity about whether users mean to use module-dependent functions or first-class module arguments. But on the other hand, what we get is that we can conflate the two notions and hide the distinction to users in most cases, instead of having to introduce a different syntax and, more problematically, having to explain to *all* users when to use which.

gasche committed:
take François Pottier's comments into account

Suggested-by: François Pottier <francois.pottier@inria.fr>

gasche commented:
I modified the PR to take @fpottier's comments into account, and updated the [rendered PDF](https://www.irif.fr/~scherer/tmp/modular-explicits-manual.pdf).

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,139 @@
-          'type' typeconstr '=' typexpr
-;
-\end{syntax}
+Mdule-dependent functions and first-class modules make it possible to
```
```suggestion
Module-dependent functions and first-class modules make it possible to
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,139 @@
-;
-\end{syntax}
+Mdule-dependent functions and first-class modules make it possible to
+manipulate modules as values, in particular to pass them as function
```
```suggestion
manipulate modules as values, in particular, to pass them as function
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,139 @@
-fewer parens required starting from 4.05)
+(First-class modules were introduced in OCaml 3.12; pattern syntax and
+package type inference introduced in 4.00; structural comparison of
+package types introduced in 4.02.; fewer parens required starting from
```
```suggestion
package types introduced in 4.02; fewer parentheses required starting from
```

Alizter replied:
```suggestion
(First-class modules were introduced in OCaml 3.12; pattern syntax and
package type inference were introduced in 4.00; structural comparison of
package types were introduced in 4.02.; fewer parens required starting from
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,139 @@
-This subset consists of named module types with optional constraints
-of a limited form: only non-parametrized types can be specified.
+Using module-dependent functions, it is also possible to express
+"sort" as a function parametrized over an ordered-set module. This
```
```suggestion
"sort" as a function parameterized over an ordered-set module. This
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -102,8 +149,10 @@ module PDF = struct let draw () = () [@@ellipsis] end
+The module expression "(val e : S)" can then \emph{unpack}
+a packed-module value "(e : module S)" back into a module. In our
+example, we can select one implementation based on command-line
+arguments:
```
```suggestion
example, one implementation can be selected based on command-line
arguments:
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,139 @@
+"sort" as a function parametrized over an ordered-set module. This
+lets callers choose their ordered-set implementation, which may be
+specialized for their type; for example sets of integers can be
+represented with bitsets, or compact representations of dense
```
```suggestion
represented with bitsets, or with compact representations of dense
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +170,127 @@ let draw_using_device device_name picture =
+(the "...") \emph{depends} on the module argument "A", then this
+function is only allowed to take static module arguments of the form
+"(module M : S)", rather than arbitrary expressions of type
+"(module S)". There is no such restriction if the the type of the body
```
```suggestion
"(module S)". There is no such restriction if the type of the body
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +170,127 @@ let draw_using_device device_name picture =
 \end{caml_example}
 
+Note that removing dependencies using "(type a)" like this can only be
+done for non-parameterized type like "elt", not with parameterized
```
```suggestion
done for non-parameterized types like "elt", not with parameterized
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,139 @@
+\end{caml_example}
+
+This provides a lightweight alternative to functors,
+indeed, this "sort" function could also be expressed as follows:
```
```suggestion
This provides a lightweight alternative to functors.
Indeed, the "sort" function could also be expressed as follows:
```

gasche commented:
Thanks! Rebased.

gasche commented:
Now that modular-explicits ( #14048 ) were merged, I will rebase this PR so that it is ready for merging as well.

(I note that no one approved it yet, which is a shame ;-)

Alizter commented:
@gasche I would like to do another pass tonight if there is no rush. It should be easier now that we have the implementation merged.

gasche commented:
There is no rush, thanks!

gasche committed:
Changes

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,139 @@
-It is not allowed in toplevel let bindings.
-Again @package-type@ can be omitted if it can be inferred from the
-enclosing pattern.
+let example = sort0 Int.compare [3; 1; 2]
```
```suggestion
let sorted = sort0 Int.compare [3; 1; 2]
```

Alizter replied:
Might as well be descriptive with the names and not repeat.

gasche replied:
(I decided to keep `example` to clarify the point that this is not meant to be a useful definition for later, and it's possibly a bit silly, just an usage example.)

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,139 @@
+\end{caml_example*}
+
+\begin{caml_example}{toplevel}
+let example = sort (module ISetIncr) [3; 1; 2];;
+let example = sort (module ISetDecr) [3; 1; 2];;
```
```suggestion
let sorted_incr = sort (module ISetIncr) [3; 1; 2];;
let sorted_decr = sort (module ISetDecr) [3; 1; 2];;
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,139 @@
+
+\lparagraph{p:fst-mod-intro-algebraic-structures}{Use-case: abstracting over algebraic structures}
+
+Module-dependent functions offer a nice way to implement code that is parametric over families of type structures, for example on any monad.
```
```suggestion
Module-dependent functions offer an effective way to implement code that is parametric over families of type structures, for example on any monad.
```

Alizter replied:
```
The sorting example uses a module with a single type. 
Module-dependent functions are especially useful with 
richer signatures, such as monads:
```

This might be a better transition sentence between examples.

gasche replied:
What I have in mind in this paragraph is that modular explicits can be used for the same sort of things as Haskell type-classes, bringing some additional structure over a (possibly-parametrized) type. I think that explicitly relating to the previous example, which is of different nature, would not help get this point across. (I could also explicitly mention type-classes and `Monad m => ...`, but this would be fairly different from the rest of the manual which is rather self-contained, and it begs the question of modular implicits, and I don't want to get into this territory for now.)

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +170,127 @@ let draw_using_device device_name picture =
+let wrong = Fun.id sort (module ISetIncr) [2; 3]
+\end{caml_example}
+
+The problem comes from subtle details in the type-checking of n-ary
+function applications: this is checked as "Fun.id" to which three
+parameters are passed, and only the type of "Fun.id" (not
+its arguments) is used to resolve the module application, so no
+module-dependent function type is known at that point.
```
This sentence is a bit of a mouthful. What about breaking the sentence at `:` instead.

Something like:

The problem arises due to the subtle details in the type-checking of n-ary function applications. This is checked as...

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +170,127 @@ let draw_using_device device_name picture =
+\lparagraph{p:fst-mod-compatibility}{Compatibility}
+
+Before OCaml 5.5, module-dependent functions were not supported. One
+could use first-class modules, but all module-parameterized functions
```
```suggestion
could use first-class modules, but all module-dependent functions
```

gasche replied:
I don't like "module-dependent functions had to be non-dependent", but I rephrased the sentence anyway. Thanks!

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -231,3 +378,83 @@ let rec to_string: 'a. 'a Typ.typ -> 'a -> string =
+When typechecking we also consider
+@'(' 'module' module-name ':' package-type ')' '->' typexpr@ and
+@'(' 'module' package-type ')' '->' typexpr@ to be identical when @module-name@
+does not appear in @typexpr@
```
```suggestion
does not appear in @typexpr@.
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,139 @@
+In the function application "sort (module ISetIncr)" from our example
+above, we say that the module argument is \emph{static}: it is known
+at compile-time.
+
```
What about adding why this distinction is important somewhere here:

```
This distinction matters because only static arguments 
allow the result type to mention module components.
```

gasche replied:
I initially added your suggestion, then added a forward-pointer to the "Limitation: selecting a module at runtime" subsection that follows, then decided that the half-explanation and forward pointer added cognitive work without bringing much and so I'm back at the current formulation...

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +170,127 @@ let draw_using_device device_name picture =
+"(module M : S)", rather than arbitrary expressions of type
+"(module S)". There is no such restriction if the type of the body
+does not mention "A", or if the type-checker can easily rewrite it to
+expand away the uses of "A".
```
alternative wording:
```
can simplify it by inlining type equalities (e.g., when A.t equals int).
```

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -231,3 +378,83 @@ let rec to_string: 'a. 'a Typ.typ -> 'a -> string =
+package with type @package-type@ and binds it to @module-name@.
+It is not allowed in toplevel let bindings. When a function takes an argument
+whose pattern is exactly @'(' 'module' module-name ':' package-type ')'@, then
+this function can be a module-dependent function.
```
Alternative wording, since I found this a bit difficult to parse:
```
A function whose parameter has the exact pattern 
(module M : S) can be module-dependent.
```
Or maybe I missed the intention here?

Alizter commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +170,127 @@ let draw_using_device device_name picture =
+  Device.draw picture
+\end{caml_example*}
+
+\lparagraph{p:fst-mod-intro-limitation-dynamic-arg}{Limitation: selecting a module at runtime}
```
I don't know if its worth elaborating what a `DEVICE` might actually look like? Or perhaps linking to some real examples.

gasche replied:
There are (mock) declarations of `module SVG : DEVICE` and `module PDF : DEVICE` earlier on in the chapter, is this what you had in mind?

Alizter reviewed (approved):
Here are some things that stuck out in my read. They are suggestions and not directions. The rest looks good content-wise, but the style is a little dry / academic. But that's not something that was started here.

gasche committed:
take Ali Caglayan's comments into account

diremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,137 @@
-;
-\end{syntax}
+Module-dependent functions and first-class modules make it possible to
+manipulate modules as values, in particular to pass them as function
```
I would say to `pass them as arguments to module-dependent functions.`. Otherwise, we should say  to receive them as function parameters. For me a function parameter is what appears in a function pattern while an argument is what appears in an application.   The intend in the end of the sentence is to say that we may pass modules as first class arguments, not that we may define module-dependent functions.

gasche replied:
I used "in particular to accept modules as function parameters.". I don't want to talk about module-dependent functions yet, this is the intro of the whole chapter and the term has not been introduced yet.

diremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,137 @@
+\end{caml_example}
+
+This provides a lightweight alternative to functors.
+Indeed, the "sort" function could also be expressed as follows:
```
```suggestion
Indeed, the "sort" function could also be expressed using a "Sort" _functor_ instead of a _function_, as follows: 
```

diremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,137 @@
+
+\lparagraph{p:fst-mod-intro-algebraic-structures}{Use-case: abstracting over algebraic structures}
+
+Module-dependent functions offer a nice way to implement code that is parametric over families of type structures, for example on any monad.
```
```suggestion
Module-dependent functions offer a nice way to implement code that is parametric over families of type structures, for example over any monad.
```

diremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,137 @@
+
+\lparagraph{p:fst-mod-intro-dynamic}{First-class modules}
+
+In the function application "sort (module ISetIncr)" from our example
```
```suggestion
In the application "sort (module ISetIncr)" from our example
```

gasche replied:
I prefer "function application".

diremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +168,127 @@ let draw_using_device device_name picture =
 
-To use this function, one can wrap the "Set.Make" functor:
+Notice that the module "MSet" is defined locally, so the type of the
+whole declaration cannot mention "MSet.elt", it would be
```
The end of the sentence `, it would be ill scoped` is unclear as we do not not what `it` refers to, and an `otherwise` is missing.  I would just remove it and write:
```
Notice that the module "MSet" is defined locally, so the type of the
whole declaration cannot mention "MSet.elt".
```

gasche replied:
I ended up with:

> Notice that "MSet.elt" cannot occur in the type of `okay`, as the
module "MSet" is only defined locally.

diremy commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +168,127 @@ let draw_using_device device_name picture =
+\begin{caml_example}{verbatim}
+let fixed = (Fun.id sort) (module ISetIncr) [2; 3]
+\end{caml_example}
+
```
```suggestion
Since the source of the problem is the unknown type of `Fun.id`, another more verbose but possible fix is to give it an explicit (partial) type annotation: 

\begin{caml_example}{verbatim}
let fixed = 
   (Fun.id : 'a -> (module Mset : Set.S) -> Mset.elt list -> Mset.elt list) 
     sort (module ISetIncr) [2; 3]
\end{caml_example}
```

gasche replied:
> (Explicitly annotating the type of `Fun.id` with a module-dependent
function type would also work.)

gasche committed:
take Didier Remy's comments into account

Octachron commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,137 @@
+\end{caml_example}
+
+This provides a lightweight alternative to functors. Indeed, the
+"sort" _function_ could also be expressed using a "Sort" _functor_:
```
`_function_` ⇒ \emph{function}
`_functor_` ⇒ \emph{functor]

gasche replied:
I just dropped the underscores, thanks!

Octachron commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -1,92 +1,137 @@
+First-class modules let you build \emph{dynamic} values that represent
+modules, where the choice of module depends on runtime
+computations. If "M" is a module of signature "S", then
+"(module M : S)" is an expression of type "(module S)", we say that it
```
I would mention in the next sentence, that ` (module M:S)` can be written `( (module M): (module S))` because it is not clear which variant is the most canonical one for first-class module expressions.

gasche replied:
I find this too obscure, I left as-is.

Octachron commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +168,129 @@ let draw_using_device device_name picture =
 
+Note that removing dependencies using "(type a)" like this can only be
+done for non-parameterized types like "elt", not with parameterized
+types like "'a t" -- there is no corresponding "(type 'a t)"
```
-- ⇒ ---

gasche replied:
;

Octachron commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +168,129 @@ let draw_using_device device_name picture =
+Note that removing dependencies using "(type a)" like this can only be
+done for non-parameterized types like "elt", not with parameterized
+types like "'a t" -- there is no corresponding "(type 'a t)"
+abstraction. It would not be possible for the "mapM" example earlier,
```
Possiblu
> for the earlier "mapM" example,
or
> for the  "mapM" example above,

?

Octachron commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -231,3 +378,82 @@ let rec to_string: 'a. 'a Typ.typ -> 'a -> string =
+When typechecking we also consider
+@'(' 'module' module-name ':' package-type ')' '->' typexpr@ and
+@'(' 'module' package-type ')' '->' typexpr@ to be identical when @module-name@
+does not appear in @typexpr@.
```
I would replace "does not appear" by "can be erased from" since "does not appear" could be interpreted as a syntactic condition which is not the case.

Octachron commented on manual/src/refman/extensions/firstclassmodules.etex:
```diff
@@ -121,29 +168,129 @@ let draw_using_device device_name picture =
 \end{caml_example}
 
-To use this function, one can wrap the "Set.Make" functor:
+Notice that "MSet.elt" cannot occur in the type of `okay`, as the
```
`okay` ⇒ "okay" (let's limit the number of quotations)

Octachron reviewed (approved):
Except for @samsa1's  and @Julow 's remark, I agree that the PR is ready to merge.
My other nitpicking remark can be ignored (or not).

At the very least this is a fine first version of the documentation for module-dependent functions.

gasche committed:
take Florian Angeletti's comments into account
